lib/ddtrace/correlation.rb

```diff
@@ -6,35 +6,109 @@ module Datadog
   # Contains behavior for managing correlations with tracing
   # e.g. Retrieve a correlation to the current trace for logging, etc.
   module Correlation
-    # Struct representing correlation
-    Identifier = Struct.new(:trace_id, :span_id, :env, :service, :version) do
-      def initialize(*args)
-        super
-        self.trace_id = trace_id || 0
-        self.span_id = span_id || 0
-        self.env = env || Datadog.configuration.env
-        self.service = service || Datadog.configuration.service
-        self.version = version || Datadog.configuration.version
+    # Represents current trace state with key identifiers
+    class Identifier
+      attr_reader \
+        :env,
+        :service,
+        :span_id,
+        :span_name,
+        :span_resource,
+        :span_service,
+        :span_type,
+        :trace_hostname,
+        :trace_id,
+        :trace_name,
+        :trace_origin,
+        :trace_process_id,
+        :trace_resource,
+        :trace_runtime_id,
+        :trace_sampling_priority,
+        :trace_service,
+        :version
+
+      # rubocop:disable Metrics/CyclomaticComplexity
+      # rubocop:disable Metrics/PerceivedComplexity
+      def initialize(
+        env: nil,
+        service: nil,
+        span_id: nil,
+        span_name: nil,
+        span_resource: nil,
+        span_service: nil,
+        span_type: nil,
+        trace_hostname: nil,
+        trace_id: nil,
+        trace_name: nil,
+        trace_origin: nil,
+        trace_process_id: nil,
+        trace_resource: nil,
+        trace_runtime_id: nil,
+        trace_sampling_priority: nil,
+        trace_service: nil,
+        version: nil
```

Comment:
How did you decide that all these fields were required? I was because of some "novel" examples, like "trace_hostname".

Reply:
`Correlation` is `TraceDigest` + some additional environment args. Everything prefixed with `span_*` should be a field from the `active_span`. Everything prefixed with `trace_*` should be a field from the `active_trace`.

Some of these fields have been added just because we could make them available. These are probably `trace_process_id`, `trace_hostname`, `trace_runtime_id`, `trace_sampling_priority`. If we consider them expensive or unwanted, we could remove them.

Some fields may appear as if they overlap, but there are instances where they don't. For example: `service`, `trace_service`, and `span_service` could be `billing-api` (default service), `billing-worker` (fork or sub-process), and `billing-db` (external call) respectively during a Sidekiq worker embedded in a Rails app that is doing a SQL query.

Some fields seem superfluous but are necessary for certain functions. For example, I'd like to update the profiler to use this `Correlation` to retrieve the `trace_id`, `trace_resource`, `span_id`, `span_type`, all of which are necessary to link profiles to traces. This feature I still need to implement.

Reply:
trace_process_id, trace_hostname, trace_runtime_id, trace_sampling_priority are the fields we discussed for removal.